### PR TITLE
TorchServe - KServe v2 - Examples update

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

**What this PR does / why we need it**:

Kserve v2 supports multiple protocols. The current examples in TorchServe seems to be outdated. 

- Adding utility to convert input image/text to bytes
- Updating mnist and bert examples to work with both bytes and tensor input (TorchServe + KServe v2 integration)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
TorchServe + KServe v2 - examples update
```

### CODE REVIEW

File: docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/bert_v2.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+{
+  "id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298",
+  "inputs": [{
+    "name": "4b7c7d4a-51e4-43c8-af61-04639f6ef4bc",
+    "shape": [1, 150],
+    "datatype": "INT64",
+    "data": [[101, 22950, 2038, 2988, 2006, 1996, 4610, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
+  }]
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/bert_v2.json]: you should be able to send multiple inputs e.g
```
{
  "id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298",
  "inputs": [{
    "name": "input_ids",
    "shape": [37],
    "datatype": "INT64",
    "data": [66, 108, 111, 111, 109, 98, 101, 114, 103, 32, 104, 97, 115, 32, 114, 101, 112, 111, 114, 116, 101, 100, 32, 111, 110, 32, 116, 104, 101, 32, 101, 99, 111, 110, 111, 109, 121]
  },
  {
     "name": "token_type_ids",
     "shape": [7],
     "datatype: "INT64",
     "data": [0,0,0,0,0,0]
  },
  {
     "name": "attention_masks",
     "shape": [7],
     "datatype: "INT64",
     "data": [1, 1, 1, 1, 1, 1, 1]
  }
  ]
}
```

File: docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/Transformer_handler_generalized.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
+from abc import ABC
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/Transformer_handler_generalized.py]: Can you help fix the lint issue ?
https://github.com/kserve/kserve/runs/5151960379?check_suite_focus=true

Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/Transformer_handler_generalized.py]: sure, i will fix the lint issues in this PR and reupload

**Comment by yuzisun** [comment]
@shrinath-suresh can you help squash the commits?

File: docs/samples/v1beta1/torchserve/v2/README.md [hunk 3]
```diff
@@ -30,11 +36,17 @@ MODEL_NAME=mnist
 SERVICE_HOSTNAME=$(kubectl get inferenceservice torchserve-v2 -o jsonpath='{.status.url}' | cut -d "/" -f 3)
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/README.md]:
```suggestion
SERVICE_HOSTNAME=$(kubectl get inferenceservice torchserve-mnist-v2 -o jsonpath='{.status.url}' | cut -d "/" -f 3)
```

Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/README.md]: Also missing
`MODEL_NAME=mnist`

File: docs/samples/v1beta1/torchserve/v2/bert/README.md [hunk 4]
```diff
@@ -38,7 +40,7 @@ The first step is to [determine the ingress IP and ports](../../../../../README.
 MODEL_NAME=torchserve-bert-v2
 SERVICE_HOSTNAME=$(kubectl get inferenceservice ${MODEL_NAME} -n <namespace> -o jsonpath='{.status.url}' | cut -d "/" -f 3)
 
-curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/v2/models/BERTSeqClassification/infer -d ./bert_v2.json
+curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/v2/models/BERTSeqClassification/infer -d ./sequence_classification/bytes/bert_v2.json
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/README.md]:
```suggestion
curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/v2/models/BERTSeqClassification/infer -d @./sequence_classification/bytes/bert_v2.json
```

File: docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+apiVersion: serving.kserve.io/v1beta1
+kind: InferenceService
+metadata:
+  name: "torchserve-bert-v2"
+spec:
+  predictor:
+    pytorch:
+      protocolVersion: v2
+      storageUri: gs://kfserving-examples/models/torchserve/huggingface
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml]: use `gs://kfserving-examples/models/torchserve/bert/v2`

File: docs/samples/v1beta1/torchserve/v2/bert/README.md [hunk 6]
```diff
@@ -8,12 +8,14 @@ on KServe.
...
+
+Download the preprocess script from [here](./sequence_classification/Transformer_handler_generalized_v2.py)
  
 ```bash
 torch-model-archiver --model-name BERTSeqClassification --version 1.0 \
 --serialized-file Transformer_model/pytorch_model.bin \
---handler ./Transformer_handler_generalized.py \
---extra-files "Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json"
+--handler ./Transformer_handler_generalized_v2.py \
+--extra-files "Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json,./Transformer_handler_generalized.py"
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: Do we need `Transformer_handler_generalized.py` in the extra files?

Comment (yuzisun) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: it is getting following error when loading the model
```
2022-02-13T19:32:53,378 [WARN ] W-9000-bert_1.0-stderr MODEL_LOG -   warnings.warn(
2022-02-13T19:32:55,921 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG - Backend worker process died.
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG - Traceback (most recent call last):
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/ts/model_service_worker.py", line 189, in <module>
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     worker.run_server()
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/ts/model_service_worker.py", line 161, in run_server
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     self.handle_connection(cl_socket)
2022-02-13T19:32:55,922 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/ts/model_service_worker.py", line 123, in handle_connection
2022-02-13T19:32:55,923 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     service, result, code = self.load_model(msg)
2022-02-13T19:32:55,923 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/ts/model_service_worker.py", line 95, in load_model
2022-02-13T19:32:55,923 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     service = model_loader.load(model_name, model_dir, handler, gpu,
2022-02-13T19:32:55,923 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/ts/model_loader.py", line 112, in load
2022-02-13T19:32:55,923 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     initialize_fn(service.context)
2022-02-13T19:32:55,924 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/model-server/tmp/models/805c05555a7941459cd1f6f32528b24a/Transformer_handler_generalized.py", line 52, in initialize
2022-02-13T19:32:55,924 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     self.tokenizer = AutoTokenizer.from_pretrained(self.setup_config["model_name"],do_lower_case=self.setup_config["do_lower_case"])
2022-02-13T19:32:55,924 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/transformers/models/auto/tokenization_auto.py", line 498, in from_pretrained
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     return tokenizer_class_fast.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/transformers/tokenization_utils_base.py", line 1743, in from_pretrained
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     return cls._from_pretrained(
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/transformers/tokenization_utils_base.py", line 1871, in _from_pretrained
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     tokenizer = cls(*init_inputs, **init_kwargs)
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/usr/local/lib/python3.8/dist-packages/transformers/models/bert/tokenization_bert_fast.py", line 199, in __init__
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     self.backend_tokenizer.normalizer = pre_tok_class(**pre_tok_state)
2022-02-13T19:32:55,925 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG - TypeError: Can't convert 'True' to PyBool
```

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh can you help squash the commits?

sure Dan . I will do it after addressing your comments.

File: docs/samples/v1beta1/torchserve/v2/bert/README.md (hunk 6 above)
Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: Yes , `Transformer_handler_generalized.py` is needed in extra files. `Transformer_handler_generalized_v2.py` inherits from `Transformer_handler_generalized.py`. 

Can you please try with `s3://kubeflow-dataset/kserve-0.5.3/bert` and let me know if you face any errors ?

File: docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/bert_v2.json (hunk 1 above)
Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/bert_v2.json]: Thank you very much @yuzisun . I have used this above mentioned request structure. This works well for single input. TS supports both static and dynamic batching. We need to discuss and come up with efficient way for batched inputs. May be we can have a discussion post this PR and raise another PR for updating batched inputs.

File: docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/Transformer_handler_generalized.py (hunk 2 above)
Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/sequence_classification/tensor/Transformer_handler_generalized.py]: The entire file is not needed. Inherited the handler class and changed only the `preprocess` method.

File: docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml (hunk 5 above)
Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml]: I used the same uri used in [bert.yaml](https://github.com/kserve/kserve/blob/48146646d76a133da09bf65f1c84f69df9e2a0e6/docs/samples/v1beta1/torchserve/v2/bert/bert.yaml#L9) . `bert.yaml` and `bert_gpu.yaml` will use the same uri.

Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml]: Are we going to host the bert example in `gs://kfserving-examples/models/torchserve/bert/v2`?  if so , i can update both the yamls to `gs://kfserving-examples/models/torchserve/bert/v2`

File: docs/samples/v1beta1/torchserve/v2/bert/README.md (hunk 6 above)
Comment (HamidShojanazeri) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: @yuzisun can you pls check if your transformers version is 4.11.0. Also, pls check if your setup_config is similar to [this one](https://github.com/pytorch/serve/blob/master/examples/Huggingface_Transformers/setup_config.json).

Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:18]: Yes. The [torchserve example](https://github.com/pytorch/serve/tree/master/examples/Huggingface_Transformers) works only with `transformers==4.6.0`. 

And please set the `captum_explanations` to `false` in `setup_config.json` until we get the captum part ready.

Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:23]: Added an instruction in readme file as well

Comment (yuzisun) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: tested the infer endpoint with your mar file it works. I was using the model from the original example bucket not sure if that was the issue.

File: docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml (hunk 5 above)
Comment (yuzisun) [reply docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml]: yes please use `gs://kfserving-examples/models/torchserve/bert/v2`

Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/bert_gpu.yaml]: sure done

File: docs/samples/v1beta1/torchserve/v2/bert/README.md (hunk 6 above)
Comment (shrinath-suresh) [reply docs/samples/v1beta1/torchserve/v2/bert/README.md:17]: > > @shrinath-suresh can you help squash the commits?
> 
> sure Dan . I will do it after addressing your comments.

Thanks Dan. i have squashed the commits.

**Pull request merged**